### PR TITLE
Allow `this.f.selector` to be pure.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/3494

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1014,6 +1014,30 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
+		_memberAccess.expression().annotation().type->category() == Type::Category::Function &&
+		member == "selector"
+	)
+		if (auto const* expr = dynamic_cast<MemberAccess const*>(&_memberAccess.expression()))
+			if (auto const* exprInt = dynamic_cast<Identifier const*>(&expr->expression()))
+				if (exprInt->name() == "this")
+					if (Declaration const* declaration = expr->annotation().referencedDeclaration)
+					{
+						u256 identifier;
+						if (auto const* variable = dynamic_cast<VariableDeclaration const*>(declaration))
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp:1030]: Hm I wonder if we actually want to support `.selector` on public variables. I don't think we had test cases for that before.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 2]
```diff
@@ -6904,15 +6904,7 @@ BOOST_AUTO_TEST_CASE(function_types_sig)
 	CHECK_ERROR(text, TypeError, "Member \"selector\" not found");
 	text = R"(
 		contract C {
-			function f() view external returns (bytes4) {
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:6907]: Why did we had a duplicate case here? :)

Perhaps something was lost during rebases.

Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:6907]: I think one of the cases used `var` before but we removed that and now they are identical.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1030]: shall I add a test case?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1030]: It works. I will add a test case.

**Comment by chriseth** [comment]
Updated.

**Comment by axic** [comment]
Fails on end-to-end tests.

**Pull request merged**